saw-core-what4/src/SAWCoreWhat4/What4.hs

```diff
@@ -988,9 +991,105 @@ parseUninterpreted ::
   UnintApp (SymExpr sym)      {- ^ Name of function and its arguments -} ->
   TValue (What4 sym)          {- ^ Type of the function's result -} ->
   IO (SValue sym)             {- ^ The symbolic value for the call to the uninterpreted function -}
-parseUninterpreted sym ref app ty =
-  evalStateT (parseUninterpreted' sym ref app ty) MapF.empty
- 
+parseUninterpreted sym ref app@(UnintApp nm args tys) ty =
+  do
+    un <- MapF.traverseWithKey mkUnint (countUninterpreted 1 MapF.empty ty)
+    evalStateT (parseUninterpreted' sym ref app ty) un)
+  where
+  mkUnint :: BaseTypeRepr tc -> UnintCount tc -> IO (Arr sym tc)
+  mkUnint ret (UnintCount n)
+    | let lim = n - 1
+    , Just (Some w) <- someNat (width lim) =
+      case isZeroOrGT1 w of
+        Left Refl ->
+          do
+            fn <- mkSymFn sym ref (nm ++ suff ret) tys ret
+            el <- W.applySymFn sym fn args
+            pure (Arr [el])
+        Right LeqProof ->
+          do
+            fn   <- mkSymFn sym ref (nm ++ suff ret) tys (BaseArrayRepr (Ctx.singleton (BaseBVRepr w) )ret)
+            arr  <- W.applySymFn sym fn args
+            els <- forM [ 0 .. lim ] $ \i ->
+              do
+                ind <- W.bvLit sym w (BV.mkBV w (fromIntegral i))
+                W.arrayLookup sym arr (Ctx.singleton ind)
+            pure (Arr els)
+    | otherwise = panic "parseUnitnerpreted" ["`someNat` returned `Nothing` for `Natural`"]
+
+  -- | Compute the number of bits required to represent the given integer.
+  width :: Natural -> Natural
+  width x = go' 0 x
+    where
+      go s 0 = s
+      go s n = let s' = s + 1 in s' `seq` go s' (n `shiftR` 1)
+
+      go' s n
+        | n < bit 32 = go s n
+        | otherwise  = let s' = s + 32 in s' `seq` go' s' (n `shiftR` 32)
```

Comment:
Could you leave some comments about what the difference between the `go` and `go'` functions are?